numba/core/entrypoints.py

```diff
@@ -1,31 +1,47 @@
 import logging
 import warnings
 
-from pkg_resources import iter_entry_points
+from numba.core.config import PYVERSION
+
+if PYVERSION < (3, 9):
+    try:
+        import importlib_metadata
+    except ImportError as ex:
+        raise ImportError(
+            "importlib_metadata backport is required for Python version < 3.9, "
+            "try:\n"
+            "$ conda/pip install importlib_metadata"
+        ) from ex
+else:
+    from importlib import metadata as importlib_metadata
+
 
 _already_initialized = False
 logger = logging.getLogger(__name__)
 
 
 def init_all():
-    '''Execute all `numba_extensions` entry points with the name `init`
+    """Execute all `numba_extensions` entry points with the name `init`
 
     If extensions have already been initialized, this function does nothing.
-    '''
+    """
     global _already_initialized
     if _already_initialized:
         return
 
     # Must put this here to avoid extensions re-triggering initialization
     _already_initialized = True
 
-    for entry_point in iter_entry_points('numba_extensions', 'init'):
-        logger.debug('Loading extension: %s', entry_point)
-        try:
-            func = entry_point.load()
-            func()
-        except Exception as e:
-            msg = "Numba extension module '{}' failed to load due to '{}({})'."
-            warnings.warn(msg.format(entry_point.module_name, type(e).__name__,
-                                     str(e)), stacklevel=2)
-            logger.debug('Extension loading failed for: %s', entry_point)
+    for entry_point in importlib_metadata.entry_points().get(
+        "numba_extensions", ()
+    ):
+        if entry_point.name == "init":
```

Comment:
I think this should be:
```suggestion
    for entry_point in importlib_metadata.entry_points(group="numba_extensions", name="init"):
```

Reply:
Thanks @eric-wieser, seems like that could work. Were this change made then I think there would also need to be a minimum version of `importlib_metadata` too (I've checked a few, earliest with `groups=` seems to be somewhere in the 3.x series, potentially 3.6 https://importlib-metadata.readthedocs.io/en/latest/history.html#v3-6-0)?
```pycon
>>> import importlib_metadata as imd
>>> imd.entry_points(group='foo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: entry_points() got an unexpected keyword argument 'group'
>>> imd.__version__
'1.6.1
```
Assuming v3.6.0 is correct, its release date is February 2021, I wonder if this is too new for a non-optional dependency with respect to some of the places with longer release/dependency cycles, e.g. HPC systems with internal curated package mirrors and linux distros.

Also, as a cross-check from the 3.10 docs: https://docs.python.org/3.10/library/importlib.metadata.html
> Changed in version 3.10: importlib.metadata is no longer provisional.

and here: https://docs.python.org/3.10/library/importlib.metadata.html#entry-points
notes that kwargs to `entry_points` are passed through to `select`.

Am going to raise this at the next public meeting to try and get a resolution: https://numba.discourse.group/t/weekly-public-meeting-every-tuesday-for-2021/658/2

Reply:
Perhaps if compatibility is a concern, we should just do feature detection on `importlib_metadata`; we don't actually need the newest version, its just the laziest way to supress the warnings.

Reply:
I think something a bit like this might work:
```python
def init_all():
    """Execute all `numba_extensions` entry points with the name `init`

    If extensions have already been initialized, this function does nothing.
    """
    global _already_initialized
    if _already_initialized:
        return

    # Must put this here to avoid extensions re-triggering initialization
    _already_initialized = True

    def load_ep(entry_point):
        """Loads a given entry point. Warns and logs on failure.
        """
        logger.debug('Loading extension: %s', entry_point)
        try:
            func = entry_point.load()
            func()
        except Exception as e:
            msg = (f"Numba extension module '{entry_point.module}' "
                   f"failed to load due to '{type(e).__name__}({str(e)})'.")
            warnings.warn(msg, stacklevel=3)
            logger.debug('Extension loading failed for: %s', entry_point)

    eps = importlib_metadata.entry_points()
    # Split, Python 3.10+ and importlib_metadata 3.6+ have the "selectable"
    # interface, versions prior to that do not. See "compatibility note" in:
    # https://docs.python.org/3.10/library/importlib.metadata.html#entry-points
    if hasattr(eps, 'select'):
        for entry_point in eps.select(group="numba_extensions", name="init"):
            load_ep(entry_point)
    else:
        for entry_point in eps.get("numba_extensions", ()):
            if entry_point.name == "init":
                load_ep(entry_point)
```
CC @gmarkall who was interested in carrying this PR to completion.

Reply:
Thanks for this - I've applied and tested it locally with Python 3.7 - 3.10 and importlib_metadata 4.10.1 - it all seemed fine locally so I've pushed it along with a merge / conflict resolution from master.